packages/components/src/custom-select-control-v2/legacy-component/index.tsx

```diff
@@ -12,18 +12,48 @@ import CustomSelectItem from '../item';
 import type { LegacyCustomSelectProps } from '../types';
 import * as Styled from '../styles';
 
+function useDeprecatedProps( {
+	showSelectedHint = false,
+	options,
+	// Deprecated
+	__experimentalShowSelectedHint,
+	...otherProps
+}: LegacyCustomSelectProps ) {
+	return {
+		...otherProps,
+		options: options.map( ( o ) => {
+			const toReturn = o;
+
+			const hint =
+				typeof o.__experimentalHint !== 'undefined'
+					? o.__experimentalHint
+					: o.hint;
+
+			if ( hint ) {
+				toReturn.hint = hint;
+			}
+
+			return toReturn;
```

Comment:
Adding the new `hint` property to each option only if necessary, because otherwise this could break consumers of the component that rely on object identity (eg. [`FontSizePicker`](https://github.com/WordPress/gutenberg/blob/9c110ce524ec56062e536bf5f3eea2a11ffdbfb6/packages/components/src/font-size-picker/font-size-picker-select.tsx#L88))

Reply:
Hmm, that's actually a breaking change for third-party consumers that could also be relying on object identity.

I'm thinking, how about we leave the options array alone, and do the `__experimental` adapting at [read time](https://github.com/WordPress/gutenberg/blob/28e9e414e7a4c6e9c383a353cd7b2177fe04135c/packages/components/src/custom-select-control-v2/legacy-component/index.tsx#L63)?